src/plugins/intel_npu/src/plugin/npuw/partitioning/online/snapshot.cpp

```diff
@@ -352,6 +400,241 @@ void Snapshot::identifyUniques() {
+
+    auto graph = getGraph();
+    for (const auto& nh : graph->sorted()) {
+        if (m_graph->contains(nh)) {
+            Group::GPtr group = m_graph->meta(nh).get<Group::GPtr>();
+            auto rep = group->repeated();
+
+            detail::GPtrSet repeating_groups;
+
+            // Note: openForMerge is not used here
+            if (rep && !group->isFrozen() && merged_this_time.count(rep) == 0) {
+                for (const auto& nh_other : graph->sorted()) {
+                    if (m_graph->contains(nh_other)) {
+                        Group::GPtr group_other = m_graph->meta(nh_other).get<Group::GPtr>();
+                        auto rep_other = group_other->repeated();
+
+                        if (rep_other && !group_other->isFrozen() && (rep_other.get() == rep.get())) {
+                            repeating_groups.insert(group_other);
+                        }
+                    }
+                }
+            }
+
+            if (!repeating_groups.empty()) {
+                auto new_rep = tryMergeTriangles(repeating_groups);
+                if (new_rep) {
+                    merged_this_time.insert(new_rep);
+                }
+            }
+        }
+    }
+
+    LOG_INFO("Number of groups after compiler pass: " << graphSize());
+    LOG_INFO("DONE");
+}
+
+// FIXME: At this point, it is almost a full duplicate of tryGrowRepeatingGroups
+std::shared_ptr<Repeated> Snapshot::tryMergeTriangles(const detail::GPtrSet& repeating_groups) {
+    auto this_rep_tag = (*(repeating_groups.begin()))->repeated();  // should be the same for each group inside
+    const auto& this_avoided = (*(repeating_groups.begin()))->avoidedTargets();
+    const auto& this_special = (*(repeating_groups.begin()))->specialTags();
+
+    if (repeating_groups.size() < 2) {
+        return nullptr;
+    }
+
+    std::unordered_map<std::vector<MetaInterconnect>, std::unordered_map<Group::GPtr, std::unordered_set<Group::GPtr>>>
+        mics;
+
+    std::vector<Group::GPtr> repeating_groups_sorted(repeating_groups.begin(), repeating_groups.end());
+
+    // FIXME: this was introduced to make the partitioning
+    // the same every run when created the same way.
+    // This std::sort allows to prioritize the groups from the tail
+    // of the original model. It's possible due to preservation of
+    // group IDs in topological order throughout the whole partitioning process.
+    // In the networks we're looking at, ensuring the merge order from the bottom
+    // of the network gives a better generalization for the identified repeated blocks,
+    // e.g. we can guarantee we can find one more, which otherwise would fuse into
+    // head or tail (depending on the topology).
+    std::sort(repeating_groups_sorted.begin(),
+              repeating_groups_sorted.end(),
+              [&](const Group::GPtr& gptr_a, const Group::GPtr& gptr_b) {
+                  return gptr_a->getId() > gptr_b->getId();
+              });
```

Comment:
I'd keep it just in case with an additional fixme comment for future. It's part of stability feature and I don't think we should spend additional time validating it here. Also, `mergeTriangles` happens in the end where we have ~200 groups instead of initial 3000+